app/src/main/java/com/example/rocketreserver/MainActivity.kt

```diff
@@ -1,39 +1,107 @@
+import androidx.compose.material.MaterialTheme
+import androidx.compose.material.Text
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.layout.ContentScale
+import androidx.compose.ui.text.font.FontWeight
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
+import coil.compose.AsyncImage
 
 class MainActivity : AppCompatActivity() {
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+        setContent { LaunchList() }
+    }
+}
 
-        setContentView(R.layout.activity_main)
+sealed class UiState {
+    object Loading : UiState()
+    object Error : UiState()
+    class Success(val launchList: List<LaunchListBlogQuery.Launch>) : UiState()
+}
 
-        lifecycleScope.launch {
-            apolloClient(this@MainActivity).subscription(TripsBookedSubscription()).toFlow()
-                .retryWhen { _, attempt ->
-                    delay(attempt * 1000)
-                    true
-                }
-                .collect {
-                    val text = when (val trips = it.data?.tripsBooked) {
-                        null -> getString(R.string.subscriptionError)
-                        -1 -> getString(R.string.tripCancelled)
-                        else -> getString(R.string.tripBooked, trips)
-                    }
-                    Snackbar.make(
-                        findViewById(R.id.main_frame_layout),
-                        text,
-                        Snackbar.LENGTH_LONG
-                    ).show()
+@Composable
+fun LaunchList() {
+
+    val context = LocalContext.current
+    // tell Compose to remember our state across recompositions
+    val state = remember {
+        apolloClient(context).query(LaunchListBlogQuery()).watch()
+            .map {
+                val launchList = it
+                    .data
+                    ?.launchConnection
+                    ?.launches
+                    ?.filterNotNull()
+                if (launchList == null) {
+                    // There were some error
+                    // TODO: do something with response.errors
+                    UiState.Error
+                } else {
+                    UiState.Success(launchList)
                 }
-        }
+            }
+            .catch { e ->
+                emit(UiState.Error)
+            }
+    }
+// collectAsState will turn our flow into State that can be consumed by Composables
+        .collectAsState(initial = UiState.Loading)
+
+// Display the com.example.rocketreserver.UiState as usual
+    when (val value = state.value) {
+        is UiState.Success -> LazyColumn(content = {
+            items(value.launchList) {
+                LaunchItem(it)
+            }
+        })
+        else -> {}
     }
 }
+
+@Composable
+fun LaunchItem(launch: LaunchListBlogQuery.Launch) {
+    Row(){AsyncImage(model = launch.mission?.missionPatch, contentDescription = null, contentScale = ContentScale.Fit )
```

Comment:
@martinbonnin 👋  I pushed the changes we worked on today as well as figured out how to apply some styling to the composable `LaunchItem`. I'm noticing a single image is oversized, and I wasn't able to get the [content scaling](https://developer.android.com/reference/kotlin/androidx/compose/ui/layout/ContentScale.Companion) to work.


perhaps you have some ideas on how to ensure that no matter how large the image is, we'll size it correctly in the Row. 


![Screen Shot 2022-08-30 at 6 32 19 PM](https://user-images.githubusercontent.com/1395158/187567612-85dd4567-50e9-4d1c-aff3-f44373befab0.png)

Reply:
You can do this with [modifiers](https://developer.android.com/jetpack/compose/modifiers):

```kotlin
        AsyncImage(
            modifier = Modifier.width(80.dp).height(80.dp),
            model = launch.mission?.missionPatch, contentDescription = null,
            contentScale = ContentScale.Fit
        )

```

Reply:
`dp` is for "device independent pixel". It's a unit that is similar to inches or millimeters and will display the same on all devices. 

You can also add padding with modifiers:

```kotlin
            modifier = Modifier.width(100.dp).height(100.dp).padding(10.dp),
```